### PR TITLE
[CALCITE-4895] MAP type in UDF cannot be externalized from json correctly

### PR DESCRIPTION
**Pull request by chunweilei**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rel/externalize/RelJson.java [hunk 1]
```diff
@@ -296,6 +296,13 @@ private RelDataType getRelDataType(RelDataTypeFactory typeFactory, Map<String, O
       componentType = toType(typeFactory, component);
       return typeFactory.createArrayType(componentType, -1);
 
+    case MAP:
+      Object key = requireNonNull(map.get("key"), "key");
+      Object value = requireNonNull(map.get("value"), "key");
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/rel/externalize/RelJson.java:301]:
```suggestion
      Object key = get(map, "key");
      Object value = get(map, "key");
```

Comment (chunweilei) [reply core/src/main/java/org/apache/calcite/rel/externalize/RelJson.java]: Thanks for your view. All comments are addressed.

File: core/src/main/java/org/apache/calcite/rel/externalize/RelJson.java [hunk 2]
```diff
@@ -387,6 +394,12 @@ private Object toJson(RelDataType node) {
       if (node.getComponentType() != null) {
         map.put("component", toJson(node.getComponentType()));
       }
+      if (node.getKeyType() != null) {
+        map.put("key", toJson(requireNonNull(node.getKeyType(), "key type")));
+      }
+      if (node.getValueType() != null) {
+        map.put("value", toJson(requireNonNull(node.getValueType(), "value type")));
+      }
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/rel/externalize/RelJson.java:402]: `requireNonNull` would not be required if you save `getKeyType()` result to a variable

Comment (chunweilei) [reply core/src/main/java/org/apache/calcite/rel/externalize/RelJson.java]: Thank you for your kindly remind. It helps a lot.

**Pull request closed without merging**